cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -454,6 +472,46 @@ func generateMountsCommonInstallRecover(mst *initramfsMountsState) error {
 	return sysconfig.ConfigureTargetSystem(configOpts)
 }
 
+func maybeMountSave(disk disks.Disk, encrypted bool, mountOpts *systemdMountOptions) (haveSave bool, err error) {
+	var saveDevice string
+	saveKey := filepath.Join(boot.InitramfsEncryptionKeyDir, "ubuntu-save.key")
+	// if ubuntu-save exists and is encrypted, the key has been created during install
+	haveSaveKey := osutil.FileExists(saveKey)
+
+	if encrypted != haveSaveKey {
```

Comment:
> ubuntu-save.key file to ubuntu-seed

ubuntu-save.key is in data, is not on seed. We get it from there after we have opened data

Reply:
I guess I still think this logic is wrong regardless, as even if ubuntu-save.key is coming from ubuntu-data, the attack I described where that file is created can still be executed, when I think we should be robust against that since we can. Granted we don't care as much in the unencrypted case, but it's still simple enough to just not care about ubuntu-save at all in the unencrypted case. 

Or are we anticipating using ubuntu-save for things even when we don't have encryption?

Reply:
remember that ubuntu-save is meant to survive and be used across reinstalls, while ubuntu-data is deleted across reinstalls, so even if there is no encryption there are still reasons to set one up

Reply:
but these codepaths are for recover and run mode, not for install mode, so I guess I don't see how that applies, unless (and I apologize for asking the same question again) we anticipate using ubuntu-save when we don't have encryption.

Another orthogonal question I guess is if a device could ever validly transition from having encryption to not having encryption? My understanding was no, so either :

1. a device never has encryption, never creates ubuntu-data encrypted, and never creates an ubuntu-save encrypted (unclear whether ubuntu-save is even created to begin with to me in this case)
1. a device always has encryption, meaning that ubuntu-save is created with encryption and never exists without encryption, and all possible run mode / recover mode boots will need to unlock ubuntu-save, and will also have an ubuntu-data-enc partition on the same disk, etc. meaning that `isDecryptDev` will always be true

I guess unless I misread this PR entirely and actually we will also be mounting ubuntu-save in install mode too? I guess thinking about that now it makes more sense to eventually mount/unlock ubuntu-save in install mode so we can use it's key(?), but this PR is not unlocking/mounting ubuntu-save in install mode AFAICT

Reply:
ubuntu-save will contain data that is used also in run mode, so we always mount it if it exist, and as I said it can exist also for uncrypted devices to preserve that data across a reinstall.

What is true is that if data is unencrypted we don't care a ton if the key file exists or not, OTOH there's a quite a few ways to brick a device both unencrypted and encrypted by messing with the encrypted file system.

Reply:
We don't have yet full reinstall (aka factory reset) code, that requires changes both here and devicestate.

Reply:
It is my understanding, and per latest agreements, encrypted ubuntu-data implies encrypted ubuntu-save, thus ubuntu-save.key file must exist (we don't really have any other way to unlock it atm).

The opposite case, unencrypted ubuntu-data does not imply whether ubuntu-save exists or not. If one exists it is mounted. I did assume that in this scenario presence of ubuntu-save.key at a well known location is unexpected and indicates that something is off. In which case, the current code errors out.

Reply:
Sorry I guess this is another case of the implemented design not being shared with me before reviewing it and so I reviewed it as I thought it should be and not how it was actually designed. 

I guess my question now is why we don't require ubuntu-save even on unencrypted devices? It seems like it would simplify the logic here and be less ambiguous as to the expected behavior. It would mean that existing beta installs are no longer valid but aren't we already claiming that beta devices are not upgradable and could be broken?

Reply:
BTW. I can relax the check and only require ubuntu-save.key presence when it's really neded, i.e. when using encryption.